drivers/timer/rk3568/timer.c

```diff
@@ -0,0 +1,207 @@
+    load_values |= (uint64_t)timestamp_timer->current_value1 << 32;
+    uint64_t ticks = UINT64_MAX - load_values;
+
+    /* convert from ticks to nanoseconds */
+    uint64_t value_whole_seconds = ticks / RK3568_TIMER_TICKS_PER_SECOND;
+    uint64_t value_subsecond_ticks = ticks % RK3568_TIMER_TICKS_PER_SECOND;
+    uint64_t value_subsecond_ns = (value_subsecond_ticks * NS_IN_S) / RK3568_TIMER_TICKS_PER_SECOND;
+    uint64_t value_ns = value_whole_seconds * NS_IN_S + value_subsecond_ns;
+
+    //sddf_dprintf("TIMER DRIVER|LOG: get_ticks_in_ns load_values: %lu value_ns: %lu\n", load_values, value_ns);
+    return value_ns;
+}
+
+void set_timeout(uint64_t ns)
+{
+    //sddf_dprintf("TIMER DRIVER|LOG: TIMEOUT SET_TIMEOUT\n");
+    //print_regs(timeout_timer);
+    /* load the timeout timer with ticks to count down from */
+    uint64_t ticks_whole_seconds = (ns / NS_IN_S) * RK3568_TIMER_TICKS_PER_SECOND;
+    uint64_t ticks_remainder = (ns % NS_IN_S) * RK3568_TIMER_TICKS_PER_SECOND / NS_IN_S ;
+    uint64_t num_ticks = ticks_whole_seconds + ticks_remainder;
+    uint32_t timeout_ticks_l = num_ticks & 0x00000000ffffffff;
+    uint32_t timeout_ticks_h = num_ticks & 0xffffffff00000000;
+
+    timeout_timer->control_reg = 0x0;
+
+    timeout_timer->load_count0 = timeout_ticks_l;
+    timeout_timer->load_count1 = timeout_ticks_h;
+
+    timeout_timer->control_reg = (RK3568_TIMER_CONTROL_TIMER_ENABLE |
+                                  RK3568_TIMER_CONTROL_MODE_USER |
+                                  RK3568_TIMER_CONTROL_INTERRUPT_ENABLE);
+    //sddf_dprintf("TIMER DRIVER|LOG: set_timeout timeout_ns: %lu ticks: %lu\n", ns, num_ticks);
+    //print_regs(timeout_timer);
+    return;
+}
+
+static void process_timeouts(uint64_t curr_time)
+{
+    for (int i = 0; i < MAX_TIMEOUTS; i++) {
+        if (timeouts[i] <= curr_time) {
+            sddf_notify(CLIENT_CH_START + i);
+            timeouts[i] = UINT64_MAX;
+        }
+    }
+
+    uint64_t next_timeout = UINT64_MAX;
+    for (int i = 0; i < MAX_TIMEOUTS; i++) {
+        if (timeouts[i] < next_timeout) {
+        //sddf_dprintf("TIMER DRIVER|LOG: next timeout at %lu i=%d\n", timeouts[i], i);
+            next_timeout = timeouts[i];
+        }
+    }
+
+    if (next_timeout != UINT64_MAX) {
+        uint64_t ns = next_timeout - curr_time;
+        set_timeout(ns);
+    }
+
+    return;
+}
+
+void init()
+{
+    assert(device_resources_check_magic(&device_resources));
+    assert(device_resources.num_irqs == 2);
+    assert(device_resources.num_regions == 1);
+
+    //sddf_dprintf("Touch mem region phys 0x%lx vaddr: %p\n", device_resources.regions[0].io_addr, (uint64_t*)device_resources.regions[0].region.vaddr);
+    //sddf_dprintf("0x%lx\n", *((uint64_t*)device_resources.regions[0].region.vaddr));
+
+    /* Ack any IRQs that were delivered before the driver started. */
+    for (int i = 0; i < device_resources.num_irqs; i++) {
+        sddf_irq_ack(device_resources.irqs[i].id);
```

Comment:
Is this necessary? Just checking since we don't generally do this, and I would be interested to know when it should happen

Reply:
Hm okay. This probably was worthy of an sDDF issue too. Is there any reason why it's only happening in some of the drivers and not all of them?